File: solutions/0470_Trapping_Rain_Water_II/priority_queue.java
class Solution {
    public int trapRainWater(int[][] heightMap) {
        int m = heightMap.length;
        int n = (m == 0 ? 0 : heightMap[0].length);
        PriorityQueue<Node> q = new PriorityQueue<>((a, b) -> a.h - b.h);
        int dirs[][] = {{0, -1}, {1, 0}, {0, 1}, {-1, 0}};
        boolean visited[][] = new boolean[m][n];

        for (int i = 0; i < n; ++i) {
            q.add(new Node(0, i, heightMap[0][i]));
            q.add(new Node(m-1, i, heightMap[m-1][i]));
            visited[0][i] = visited[m-1][i] = true;
        }
        for (int i = 1; i < m-1; ++i) {
            q.add(new Node(i, 0, heightMap[i][0]));
            q.add(new Node(i, n-1, heightMap[i][n-1]));
            visited[i][0] = visited[i][n-1] = true;
        }

        int ret = 0;
        while (!q.isEmpty()) {
            Node cur = q.poll();
            for (int d = 0; d < 4; ++d) {
                int r = dirs[d][0] + cur.r, c = dirs[d][1] + cur.c;
                if (r < 0 || r >= m || c < 0 || c >= n || visited[r][c]) continue;
                ret += Math.max(0, cur.h - heightMap[r][c]);
                q.add(new Node(r, c, Math.max(cur.h, heightMap[r][c])));
                visited[r][c] = true;
            }
        }
        return ret;
    }

    class Node {
        int r;
        int c;
        int h;

        Node(int r, int c, int h) {
            this.r = r;
            this.c = c;
            this.h = h;
        }
    }
}

File: solutions/0407_Trapping_Rain_Water_II/priority_queue.py
import heapq as hq

class Solution:

    def rainCnt(self, heightMap):
        m, n = len(heightMap), len(heightMap[0])
        visited = [[False for _ in range(n)] for _ in range(m)]
        dirs = [[-1, 0], [0, -1], [1, 0], [0, 1]]
        # add all bounding items into heap
        q = []
        for i in range(m):
            q += [Node(i, 0, heightMap[i][0])]
            q += [Node(i, n - 1, heightMap[i][n - 1])]
            visited[i][0], visited[i][n - 1] = True, True
        for i in range(n):
            q += [Node(0, i, heightMap[0][i])]
            q += [Node(m - 1, i, heightMap[m - 1][i])]
            visited[0][i], visited[m - 1][i] = True, True
        hq.heapify(q)
        # poll smallest height item from heap
        ret = 0
        while q:
            cur = hq.heappop(q)
            for dir in dirs:
                r, c = cur.r + dir[0], cur.c + dir[1]
                if 0 <= r < m and 0 <= c < n and not visited[r][c]:
                    ret += max(0, cur.v - heightMap[r][c])
                    hq.heappush(q, Node(r, c, max(cur.v, heightMap[r][c])))
                    visited[r][c] = True
        return ret

class Node(object):

    def __init__(self, r, c, v):
        self.r = r
        self.c = c
        self.v = v

    def __repr__(self):
        return f'Node: {self.r, self.c ,self.v}'

    def __lt__(self, other):
        return self.v < other.v

input = [
        [3, 3, 3, 3, 3],
        [3, 2, 2, 2, 3],
        [3, 2, 1, 2, 3],
        [3, 2, 2, 2, 3],
        [3, 3, 3, 3, 3],
        ]
x = Solution().rainCnt(input)
print(x)
